src/App.tsx

```diff
@@ -1,24 +1,197 @@
+import { Notifications } from './components/Notifications';
+import { TodoList } from './components/TodoList';
+import { FilterOptions } from './types/FilterOptions';
+import { Todo } from './types/Todo';
+import {
+  getTodos, createTodo, removeTodo, updateTodo,
+} from './api/todos';
 import { UserWarning } from './UserWarning';
 
-const USER_ID = 0;
+const USER_ID = 10873;
 
 export const App: React.FC = () => {
+  const [todos, setTodos] = useState<Todo[]>([]);
+  const [error, setError] = useState<string | null>(null);
+  const [filter, setFilter] = useState<FilterOptions>(FilterOptions.ALL);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [loadingTodos, setLoadingTodos] = useState([0]);
+
+  useEffect(() => {
+    const fetchData = async () => {
+      try {
+        const todosFromServer: Todo[] = await getTodos(USER_ID);
+
+        setTodos(todosFromServer);
+      } catch {
+        setError('Unable to get todos');
+      }
+    };
+
+    fetchData();
+  }, []);
+
+  useEffect(() => {
+    let timeoutId: NodeJS.Timeout;
+
+    if (error) {
+      timeoutId = setTimeout(() => setError(null), 3000);
+    }
+
+    return () => {
+      clearTimeout(timeoutId);
+    };
+  }, [error]);
+
+  const activeTodos = todos.filter(todo => !todo.completed);
+  const completedTodos = todos.filter(todo => todo.completed);
+  const isAllTodosCompleted = todos.every(todo => todo.completed);
+
+  const visibleTodos = useMemo(() => {
+    switch (filter) {
+      case FilterOptions.ACTIVE:
+        return activeTodos;
+
+      case FilterOptions.COMPLETED:
+        return completedTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, filter]);
+
+  const addTodo = async (title: string) => {
+    try {
+      const newTodo = {
+        title,
+        userId: USER_ID,
+        completed: false,
+      };
+
+      setTempTodo({
+        id: 0,
+        ...newTodo,
+      });
+
+      const createdTodo = await createTodo(newTodo);
+
+      setTodos(prevTodos => [...prevTodos, createdTodo]);
+    } catch {
+      setError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+    }
+  };
+
+  const updateTodoInfo = async (
+    todoId: number,
+    newTodoData: Partial<Pick<Todo, 'title' | 'completed'>>,
+  ) => {
+    try {
+      setLoadingTodos(prevIds => [...prevIds, todoId]);
+      const updatedTodo = await updateTodo(todoId, newTodoData);
+
+      setTodos(prevTodos => prevTodos.map(todo => {
+        if (todo.id !== todoId) {
+          return todo;
+        }
+
+        return updatedTodo;
+      }));
+    } catch {
+      setError('Unable to update a todo');
+    } finally {
+      setLoadingTodos(prevIds => prevIds.filter(id => id !== todoId));
+    }
+  };
+
+  const handleToggleAllButton = () => {
+    todos.forEach(async (todo) => {
+      await updateTodoInfo(
+        todo.id,
+        { completed: !isAllTodosCompleted },
+      );
+    });
+  };
```

Comment:
Before sending a request, You should check if you have a completed todo, and if there is, send only those that are not completed for a change to avoid redundant requests to the server.